querier.cc

```diff
@@ -321,6 +334,83 @@ void querier_cache::insert_shard_querier(query_id key, shard_mutation_querier&&
+        const dht::partition_range& range,
+        const query::partition_slice& slice,
+        reader_concurrency_semaphore& current_sem,
+        tracing::trace_state_ptr trace_state,
+        db::timeout_clock::time_point timeout) {
+    auto base_ptr = find_querier(index, key, range, trace_state);
+    auto& stats = _stats;
+    ++stats.lookups;
+    if (!base_ptr) {
+        ++stats.misses;
+        return std::nullopt;
+    }
+
+    auto* q_ptr = dynamic_cast<Querier*>(base_ptr.get());
+    if (!q_ptr) {
+        throw std::runtime_error("lookup_querier(): found querier is not of the expected type");
+    }
+    auto& q = *q_ptr;
+    auto reader_opt = q.permit().semaphore().unregister_inactive_read(querier_utils::get_inactive_read_handle(q));
+    if (!reader_opt) {
+        throw std::runtime_error("lookup_querier(): found querier that is evicted");
+    }
+    reader_opt->set_timeout(timeout);
+    querier_utils::set_reader(q, std::move(*reader_opt));
+    --stats.population;
+
+    const auto can_be_used = can_be_used_for_page(_is_user_semaphore_func, q, s, range, slice, current_sem);
+    if (can_be_used == can_use::yes) {
+        tracing::trace(trace_state, "Reusing querier");
+        return std::optional<Querier>(std::move(q));
+    }
+
+    tracing::trace(trace_state, "Dropping querier because {}", cannot_use_reason(can_be_used));
+    ++stats.drops;
+
+    auto permit = q.permit();
+
+    // Save semaphore name and address for later to use it in
+    // error/warning message
+    auto q_semaphore_name = permit.semaphore().name();
+    auto q_semaphore_address = reinterpret_cast<uintptr_t>(&permit.semaphore());
+
+    // Close and drop the querier in the background.
+    // It is safe to do so, since _closing_gate is closed and
+    // waited on in querier_cache::stop()
+    (void)with_gate(_closing_gate, [q = std::move(q)] () mutable {
+        return q.close().finally([q = std::move(q)] {});
+    });
+
+    if (can_be_used == can_use::no_scheduling_group_mismatch) {
+        ++stats.scheduling_group_mismatches;
+        qlogger.warn("user semaphore mismatch detected, dropping reader {}: "
+                "reader belongs to {} (0x{:x}) but the query class appropriate is {} (0x{:x})",
+                    permit.description(),
+                    q_semaphore_name,
+                    q_semaphore_address,
+                    current_sem.name(),
+                    reinterpret_cast<uintptr_t>(&current_sem));
+    }
+    else if (can_be_used == can_use::no_fatal_semaphore_mismatch) {
+        on_internal_error(qlogger, format("semaphore mismatch detected, dropping reader {}: "
+                "reader belongs to {} (0x{:x}) but the query class appropriate is {} (0x{:x})",
+                permit.description(),
+                q_semaphore_name,
+                q_semaphore_address,
+                current_sem.name(),
+                reinterpret_cast<uintptr_t>(&current_sem)));
+    }
+
+    return std::nullopt;
+}
```

Comment:
I don't believe this complication is worth 30 instructions.

Reply:
Agree.

Reply:
makes sense